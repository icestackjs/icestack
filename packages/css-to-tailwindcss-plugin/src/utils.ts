import path from 'node:path'
import fs from 'node:fs'
export function resolvePath(p: string, cwd: string = process.cwd()) {
  return path.isAbsolute(p) ? p : path.resolve(cwd, p)
}

export function ensureDir(dir: string) {
  if (fs.existsSync(dir)) {
    const stat = fs.statSync(dir)
    if (!stat.isDirectory()) {
      fs.mkdirSync(dir)
    }
  } else {
    fs.mkdirSync(dir)
  }
}
