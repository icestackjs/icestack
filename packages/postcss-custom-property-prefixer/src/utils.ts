import { PrefixFunction, UserDefinedOptions } from './types'

export function matchCustomPropertyFromValue(str: string, cb: (arr: RegExpExecArray, index: number) => void) {
  let arr: RegExpExecArray | null
  let index = 0
  const regex = /var\(\s*(--\w[\w-]*)/g
  while ((arr = regex.exec(str)) !== null) {
    cb(arr, index)
    index++
  }
}

export function makeCustomProperty(customProperty: string, prefix: string) {
  return `--${prefix}${customProperty.slice(2)}`
}

export const postcssPlugin = 'postcss-custom-property-prefixer'

export function makePrefixFunction(prefix: UserDefinedOptions['prefix']): PrefixFunction {
  return typeof prefix === 'string'
    ? () => {
        return prefix
      }
    : prefix
}
