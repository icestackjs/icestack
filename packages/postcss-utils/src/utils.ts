import type { ChildNode } from 'postcss'
import sassParser from 'postcss-scss'
import selectorParser from 'postcss-selector-parser'
import { set, get } from 'lodash'
import type { CssValue, CssInJs } from '@icestack/types'

const defaultSelectorParser = selectorParser()

export function compressCssSelector(selectors: string) {
  return defaultSelectorParser.processSync(selectors, { lossless: false })
}

export function recursiveNodes(nodes: ChildNode[], result: Record<string, any> = {}) {
  for (const node of nodes) {
    switch (node.type) {
      case 'atrule': {
        if (node.name === 'apply') {
          const v = get(result, 'apply')
          if (Array.isArray(v)) {
            v.push(node.params)
          } else {
            set(result, 'apply', [node.params])
          }
        } else {
          const selector = `@${node.name} ${node.params}`
          result[selector] = {}
          recursiveNodes(node.nodes ?? [], result[selector])
        }

        break
      }
      case 'rule': {
        const s = compressCssSelector(node.selector)
        if (result[s] === undefined) {
          result[s] = {}
        }

        recursiveNodes(node.nodes, result[s])
        break
      }
      case 'decl': {
        set(result, `css.${node.prop}`, node.value + (node.important ? ' !important' : ''))
        break
      }
    }
  }
  return result
}

export function transformCss2Js(css: string): Record<string, any>
export function transformCss2Js(css: string | CssInJs): Record<string, any>
export function transformCss2Js<T>(css: T) {
  if (typeof css === 'string') {
    const root = sassParser.parse(css)
    const result = recursiveNodes(root.nodes)
    return result
  }
  return css
}

export function mapCss2JsArray(value?: CssValue) {
  if (value) {
    return Array.isArray(value) ? value.map((x) => transformCss2Js(x)) : [transformCss2Js(value)]
  }
  return []
}
