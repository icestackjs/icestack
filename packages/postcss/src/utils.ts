export * from '@icestack/postcss-utils'
