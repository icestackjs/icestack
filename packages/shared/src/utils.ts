import { isObject, set, get } from 'lodash'
import postcss from 'postcss'
import selectorParser from 'postcss-selector-parser'

import { createDefu } from 'defu'

const defuOverrideArray = createDefu((obj, key, value) => {
  if (Array.isArray(obj[key]) && Array.isArray(value)) {
    obj[key] = value
    return true
  }
})

const defuOverrideApplyCss = createDefu((obj, key, value) => {
  if ((key === 'css' || key === 'apply') && typeof obj[key] === 'object' && typeof value === 'object') {
    obj[key] = value
    return true
  }
})

export { defuOverrideArray, defuOverrideApplyCss }

const defaultSelectorParser = selectorParser()

export function compressCssSelector(selectors: string) {
  return defaultSelectorParser.processSync(selectors, { lossless: false })
}

export function expandTypes(types: string[], fn: (typeName: string) => { key: string; value: object }) {
  return types.reduce<Record<string, object>>((acc, cur) => {
    const { key, value } = fn(cur)
    acc[key] = value
    return acc
  }, {})
}

export function preprocessCssInJs(obj: Record<string, any>, res: Record<string, any> = {}) {
  if (typeof obj !== 'object') {
    return res
  }
  const keys = Object.keys(obj)

  for (const key of keys) {
    const value = obj[key]
    if (key === 'apply') {
      if (typeof value === 'string') {
        if (Array.isArray(res[key])) {
          res[key].push(value.split(' '))
        } else {
          res[key] = [value.split(' ')]
        }
      } else if (Array.isArray(value)) {
        // has been handled
        // preprocessCssInJs case 2

        for (const applyStr of value) {
          if (typeof applyStr === 'string') {
            if (Array.isArray(res[key])) {
              res[key].push(applyStr.split(' '))
            } else {
              res[key] = [applyStr.split(' ')]
            }
          }
        }

        // res[key] = value
      }
      // do nothing
    } else if (key === 'css') {
      res[key] = value
    } else if (isObject(value)) {
      const s = compressCssSelector(key)
      if (res[s] === undefined) {
        res[s] = {}
      }

      preprocessCssInJs(obj[key], res[s])
    }
  }
  return res
}

export function getSelector(type: string, prefix: string = '-') {
  return type === '' ? type : `${prefix}${type}`
}

export function recursiveNodes(nodes: postcss.ChildNode[], result: Record<string, any> = {}) {
  for (const node of nodes) {
    switch (node.type) {
      case 'atrule': {
        if (node.name === 'apply') {
          const v = get(result, 'apply')
          if (Array.isArray(v)) {
            v.push(node.params)
          } else {
            set(result, 'apply', [node.params])
          }
        } else {
          const selector = `@${node.name} ${node.params}`
          result[selector] = {}
          recursiveNodes(node.nodes, result[selector])
        }

        break
      }
      case 'rule': {
        const s = compressCssSelector(node.selector)
        if (result[s] === undefined) {
          result[s] = {}
        }

        recursiveNodes(node.nodes, result[s])
        break
      }
      case 'decl': {
        set(result, `css.${node.prop}`, node.value + (node.important ? ' !important' : ''))
        break
      }
    }
  }
  return result
}

export function transformCss2Js(css: string) {
  const root = postcss.parse(css)
  const result = recursiveNodes(root.nodes)
  return result
}

export { default as defu } from 'defu'
