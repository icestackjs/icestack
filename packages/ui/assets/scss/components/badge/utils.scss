@use "../../utils/index.scss" as *;
@use "sass:map";
$S: inject("badge.selector");
$DefaultMap: inject("badge.defaults.utils");

#{$S} {
  @each $typeName, $map in map.get($DefaultMap, "sizes") {
    &#{getSelector($typeName)} {
      @include resolve(map.get($map, "default"));
    }
  }
}
