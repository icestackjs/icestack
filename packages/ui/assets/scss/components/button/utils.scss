@use "../../utils/index.scss" as *;
@use "sass:map";
$S: inject("button.selector");
$DefaultMap: inject("button.defaults.utils");

#{$S} {
  @each $size, $map in map.get($DefaultMap, "sizes") {
    &#{getSelector($size)} {
      @include resolve(map.get($map, "default"));
    }
  }

  @each $shape, $map in map.get($DefaultMap, "shapes") {
    &#{getSelector($shape)} {
      @include resolve(map.get($map, "default"));

      @each $size, $m in map.get($map, "sizes") {
        &:where(#{getSelector($size,#{$S}-)}) {
          @include resolve(map.get($m, "default"));
        }
      }
    }
  }
}
