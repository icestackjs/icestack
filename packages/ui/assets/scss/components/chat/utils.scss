@use "../../utils/index.scss" as *;

@include spread(inject("chat.defaults.utils"));
