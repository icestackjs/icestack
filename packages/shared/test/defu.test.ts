import { defuOverrideArray, defu, defuArrayRight } from '@/utils'

describe('defu', () => {
  it('defuOverrideArray case 0', () => {
    const a = {
      a: {
        b: {
          c: [1, 2]
        }
      }
    }
    const b = {
      a: {
        b: {
          c: [3, 4]
        }
      }
    }
    expect(defuOverrideArray(b, a)).toEqual(b)
  })

  it('defuOverrideArray case 1', () => {
    const a = {
      a: {
        b: {
          a: [-1, 1],
          c: [1, 2]
        }
      }
    }
    const b = {
      a: {
        b: {
          c: [3, 4]
        }
      }
    }
    expect(defuOverrideArray(b, a)).toEqual({
      a: {
        b: {
          a: [-1, 1],
          c: [3, 4]
        }
      }
    })
  })

  // it('defuOverrideApplyCss case 0', () => {
  //   const a = {
  //     apply: [1, 2],
  //     css: {
  //       a: 1,
  //       b: 2
  //     }
  //   }

  //   const b = {
  //     apply: [],
  //     css: {}
  //   }

  //   expect(defuOverrideApplyCss(b, a)).toEqual(b)
  // })

  it('defu array', () => {
    const arr = defu({ array: ['b', 'c'] }, { array: ['a'] })
    expect(arr).toMatchSnapshot()
  })

  it('defuArrayRight ', () => {
    const arr = defuArrayRight({ array: ['b', 'c'] }, { array: ['a'] })
    expect(arr).toMatchSnapshot()
  })
})
