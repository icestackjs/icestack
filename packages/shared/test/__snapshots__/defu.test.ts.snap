// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defu > defu array 1`] = `
{
  "array": [
    "b",
    "c",
    "a",
  ],
}
`;

exports[`defu > defuArrayRight  1`] = `
{
  "array": [
    "a",
    "b",
    "c",
  ],
}
`;
